src/main/java/org/prebid/server/bidder/taboola/TaboolaBidder.java

```diff
@@ -0,0 +1,169 @@
+import com.iab.openrtb.request.Imp;
+import com.iab.openrtb.request.Publisher;
+import com.iab.openrtb.request.Site;
+import com.iab.openrtb.response.BidResponse;
+import com.iab.openrtb.response.SeatBid;
+import io.vertx.core.MultiMap;
+import io.vertx.core.http.HttpMethod;
+import org.apache.commons.collections4.CollectionUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.prebid.server.bidder.Bidder;
+import org.prebid.server.bidder.model.BidderBid;
+import org.prebid.server.bidder.model.BidderCall;
+import org.prebid.server.bidder.model.BidderError;
+import org.prebid.server.bidder.model.HttpRequest;
+import org.prebid.server.bidder.model.Result;
+import org.prebid.server.exception.PreBidException;
+import org.prebid.server.json.DecodeException;
+import org.prebid.server.json.JacksonMapper;
+import org.prebid.server.proto.openrtb.ext.ExtPrebid;
+import org.prebid.server.proto.openrtb.ext.request.taboola.ExtImpTaboola;
+import org.prebid.server.proto.openrtb.ext.response.BidType;
+import org.prebid.server.util.HttpUtil;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+
+public class TaboolaBidder implements Bidder<BidRequest> {
+
+    private static final String DISPLAY_ENDPOINT_PREFIX = "display";
+
+    private static final TypeReference<ExtPrebid<?, ExtImpTaboola>> TABOOLA_EXT_TYPE_REFERENCE =
+            new TypeReference<>() {
+            };
+
+    private final String endpointTemplate;
+    private final JacksonMapper mapper;
+    private final MultiMap headers;
+
+    public TaboolaBidder(String endpointTemplate, JacksonMapper mapper) {
+        this.endpointTemplate = HttpUtil.validateUrl(Objects.requireNonNull(endpointTemplate));
+        this.mapper = Objects.requireNonNull(mapper);
+        headers = HttpUtil.headers();
+    }
+
+    @Override
+    public Result<List<HttpRequest<BidRequest>>> makeHttpRequests(BidRequest request) {
+        final List<BidderError> errors = new ArrayList<>();
+        final List<HttpRequest<BidRequest>> httpRequests = new ArrayList<>();
+
+        for (Imp imp : request.getImp()) {
+            try {
+                validateImp(imp);
+
+                final ExtImpTaboola impExt = parseImpExt(imp);
+                final String type = getBidType(imp).equals(BidType.banner)
+                        ? DISPLAY_ENDPOINT_PREFIX
+                        : BidType.xNative.getName();
+
+                final BidRequest outgoingRequest = createRequest(request, imp, impExt);
+
+                httpRequests.add(HttpRequest.<BidRequest>builder()
+                        .method(HttpMethod.POST)
+                        .uri(buildEndpointUrl(impExt, type, outgoingRequest.getSite().getDomain()))
+                        .body(mapper.encodeToBytes(outgoingRequest))
+                        .headers(headers)
+                        .payload(outgoingRequest)
+                        .build());
```

Comment:
I have added all of this as a `try-catch` block due to avoid duplicate requests. Because if we get an error with an invalid imp, we can still get BidRequest after this. So, in my opinion, it should all be in `try-catch`.

Reply:
Just use `continue` keyword in this case.